File: oio337a/Algorithm_2/day_1/33.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        s, e, max_idx = 0, len(nums) - 1, -1 # 초기값 세팅
        if e == -1: # nums 가 비어있을 때
            return -1
        if nums[s] > nums[e]: # nums 가 로테이션 되어있을 때
            while s <= e: # max_idx 찾기 (이진탐색)
                max_idx = (s + e) // 2
                '''
                기준을 맨 뒤에 있는 값으로 확인
                nums = 3,7,8,0,1,2 이 상태 라면
                만약 mid 가 8 일 경우에는 마지막 의 값 2 보다 크므로 s를 오른쪽 옮길 수 있음
                만약 mid 가 0 일 경우에는 마지막 의 값 2 보다 작으므로 e를 왼쪽으로 옯길 수 있음
                '''
                if nums[max_idx] > nums[-1]:
                    s = max_idx + 1
                elif nums[max_idx] < nums[-1]:
                    e = max_idx - 1
                if nums[max_idx] > nums[max_idx + 1]:
                    break
                elif nums[max_idx] < nums[max_idx - 1]:
                    max_idx -= 1
                    break
            if target <= nums[-1]: # target을 보고 s, e 다시 할당
                s, e = max_idx + 1, len(nums) - 1
            else:
                s, e = 0, max_idx
            while s <= e:
                mid = (s + e) // 2
                if nums[mid] > target:
                    e = mid - 1
                elif nums[mid] < target:
                    s = mid + 1
                else:
                    return mid
        else: # nums 가 로테이션 안된 상태 걍 이진 탐색
            while s <= e:
                mid = (s + e) // 2
                if nums[mid] > target:
                    e = mid - 1
                elif nums[mid] < target:
                    s = mid + 1
                else:
                    return mid
        return -1

nums = [3, 1]
target = 1
test = Solution()
print(test.search(nums, target))

File: saewoo1/leet2/day1/33.java
class Solution {
    public int search(int[] nums, int target) {
        /*
			일단 로테이션이라는거 자체에서 연속된 수는 어느정도 정렬이 되어있다는 보장을 받음
			범위를 mid를 기준으로 나누고, target이 left사이 에 있는지, right 사이에 있는지 확인
			찾았다면, right, left 값을 조절해 범위를 줄이며 다시 이분탐색
		 */

        int left = 0;
        int right = nums.length - 1;
        while (left <= right)
        {
            int mid = (right + left) / 2;
            if (nums[mid] == target)
                return (mid);
            if (nums[left] <= nums[mid]) // 이 범위는 정렬이 됐구나 보장받음
            {
                if (nums[left] <= target && nums[mid] > target) // left와 mid 범위 사이에 target이 존재한다면
                    right = mid - 1; // left는 고정시키고, right의 범위를 줄인다
                else
                    left = mid + 1; // target이 그 사이값이 아니라면, left를 mid 다음으로 넘기고 재탐색
            }
            else // 반대로 nums[right] > nums[mid]로 뒷부분이 정렬이 됐다고 판별이 나면
            {
                if (nums[right] >= target && nums[mid] < target) // 이 사이에 target이 존재하는가?
                    left = mid + 1; // 존재한다면 right를 고정시키고, left 범위를 줄인다
                else
                    right = mid - 1; // 존재하지 않는다면, right의 범위를 mid 전으로 옮겨 범위를 줄인다
            } // mid를 기준으로 left, right를 옮겼으니 정렬된 부분을 찾을 수밖에 없음을 보장받읍니다
        }
        return (-1);
    }
}